lms/djangoapps/mobile_api/middleware.py

```diff
@@ -0,0 +1,78 @@
+"""
+Middleware for Mobile APIs
+"""
+from django.conf import settings
+from django.http import HttpResponse
+import re
+from mobile_api.models import MobileAppVersionConfig
+from openedx.core.lib.mobile_utils import is_request_from_mobile_app
+
+
+class AppVersionUpgrade(object):
+    """
+    Middleware class to keep track of mobile application version being used
+    """
+    LATEST_VERSION = "EDX-APP-LATEST-VERSION"
+    UPGRADE_DEADLINE = "EDX-APP-UPGRADE-DATE"
+    MOBILE_PLATFORM_USER_AGENT_REGEX = {
+        MobileAppVersionConfig.ANDROID: (r'Dalvik/[.0-9]+ \(Linux; U; Android [.0-9]+; '
+                                         r'(.*) Build/[0-9a-zA-Z]*\) (.*)/[0-9]+.[0-9]+.[0-9]+(.[0-9a-zA-Z]*)?'),
+        MobileAppVersionConfig.IOS: (r'\([0-9]+.[0-9]+.[0-9]+(.[0-9a-zA-Z]*)?; '
+                                     r'OS Version [0-9.]+ \(Build [0-9a-zA-Z]*\)\)'),
+    }
+    MOBILE_PLATFORM_VERSION = {
+        # version is found as last entry in android user agent after "/"
+        MobileAppVersionConfig.ANDROID: lambda user_agent: user_agent[user_agent.rfind("/") + 1:],
+        # version is found between first occurrence ( and ; i.e. (version;
+        MobileAppVersionConfig.IOS: lambda user_agent: user_agent[user_agent.find("(") + 1:user_agent.find(";")],
+    }
+
+    def form_factor(self, user_agent):
+        """ extracts mobile platform and app version in use """
+        form_factor = None
+        for platform, user_agent_regex in self.MOBILE_PLATFORM_USER_AGENT_REGEX.iteritems():
+            if re.search(user_agent_regex, user_agent):
+                user_app_version = self.MOBILE_PLATFORM_VERSION[platform](user_agent)
+                # if user app version does not match to X.X.X.Y (acceptable format) then let pass the request
+                # as it might belong to older apps, so return None
+                if self.validate_version(user_app_version):
+                    return {"platform": platform, "user_app_version": user_app_version}
```

Comment:
I would like to keep it argumented; either dict or named-tuple, I think namedtuple would be a better choice.